### PR TITLE
resolves #1545: Type system for Values

### LINKED ISSUE #1545: provide a type system for expressions and values
We currently plan queries in a way that we assume that the result of a query is of the record type the query defined over. That prohibits us from doing any computations that may happen along the access path of a plan. In fact, a type system is crucial for more sophisticated queries.

We would like to support:
- primitives (`INT`, `LONG`, `FLOAT`, `DOUBLE`, `STRING`, `BYTES`, `BOOLEAN`)
- structured (`ARRAY`, `TUPLE`)

Arrays can define an element type, tuples define a map of name of a field to some type of that field. Arrays and tuples can be arbitrarily nested.
Fields of tuples can be addressed through an optional field name or an ordinal.
Types are non-referential, i.e. you cannot refer to a type by name. Types compare by structure.

Types used by the semantic modeling and protobuf encoding of actual data must complementing. Types can be transformed into protobuf descriptors and back in a _lossless_ way.

Existing data on disk MUST not be affected in a sense that existing queries can derive a return type of the query (the describe type) that properly describes what is being returned. The protobuf descriptor that can be computed from the describe type MUST be able to allow for parsing of the data (without looking at the descriptor of the message directly).

The data on disk being returned today MUST not be migrated into a new shape or form in any way. Existing queries MUST return their data in an encoding that is wire-compatible with the previous format.

### PR DESCRIPTION
**Pull request by hatyo**
This introduces several constructs that help defining types on `Value`s (and later on other components of QGM). The type definitions themselves are very close to the ones in Protobuf, currently there are primitive and structure types, moreover, types employ structural equality and they can be losslessly translated to and from Protobuf descriptors.

A new interface `Typed` is added, this interface allows implementors (typically expressions), to expose `Type` information of their result sets. in this PR the scalar `Value`s implement this interface with limited support.

A number of tests are added to examine translation between `Type`s and synthetically generated Protobuf `Descriptor`s.

This resolves https://github.com/FoundationDB/fdb-record-layer/issues/1545.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/QuantifiedColumnValue.java [hunk 1]
```diff
@@ -130,6 +153,11 @@ public boolean isFunctionallyDependentOn(@Nonnull final Value otherValue) {
 
     @Nonnull
     public static QuantifiedColumnValue of(@Nonnull CorrelationIdentifier alias, int ordinal) {
-        return new QuantifiedColumnValue(alias, ordinal);
+        return new QuantifiedColumnValue(alias, ordinal, Type.Record.erased());
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/QuantifiedColumnValue.java:156]: Can you add a TODO for this one?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java [hunk 2]
```diff
@@ -54,14 +57,34 @@
...
+    default Type getResultType() {
+        return Type.primitiveType(Type.TypeCode.UNKNOWN);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nonnull
+    @Override
+    default String explain(@Nonnull final Formatter formatter) {
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java:83]: Maybe that one should be `describe()` which is more in line what other Das do.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java [hunk 3]
```diff
@@ -201,14 +224,6 @@ default Value translateLeaf(@Nonnull final Map<Value, Value> translationMap) {
         return narrowedClass.cast(this);
     }
 
-    @Nonnull
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java:204]: I wonder why this is dropped.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/temp/TypeTest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.stream.Stream;
+
+/**
+ * Tests for synthesizing a protobuf descriptor from a {@link Type} object.
+ */
+class TypeTest {
+
+    static class ProtobufRandomMessageProvider implements ArgumentsProvider {
+
+        private static final int seed = 42;
```
Comment (normen662) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/temp/TypeTest.java:53]: Ah yes!

**Review by normen662 — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Formatter.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Formatter.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Formatter.java:6]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/OrdinalFieldValue.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+import com.apple.foundationdb.record.query.predicates.ValueWithChild;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A value representing the quantifier as an object.
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/OrdinalFieldValue.java:40]: Done, I also added an example.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/SemanticException.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Semantic exceptions that could occur e.g. to illegal type conversions, ... etc.
+ */
+public class SemanticException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/SemanticException.java:28]: Right, I was thinking in the same direction, basically this exception represents errors coming mainly from incorrect queries (e.g. a `RecordType` is not found).

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java [hunk 8]
```diff
@@ -0,0 +1,1400 @@
...
+        DOUBLE(Double.class, FieldDescriptorProto.Type.TYPE_DOUBLE, true, true),
+        FLOAT(Float.class, FieldDescriptorProto.Type.TYPE_FLOAT, true, true),
+        INT(Integer.class, FieldDescriptorProto.Type.TYPE_INT32, true, true),
+        LONG(Long.class, FieldDescriptorProto.Type.TYPE_INT64, true, true),
+        STRING(String.class, FieldDescriptorProto.Type.TYPE_STRING, true, false),
+        TUPLE(List.class, null, false, false),
+        RECORD(Message.class, null, false, false),
+        ARRAY(Array.class, null, false, false),
+        STREAM(null, null, false, false),
+        FUNCTION(null, null, false, false);
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java:297]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java [hunk 9]
```diff
@@ -0,0 +1,1400 @@
...
+        @Override
+        public String toString() {
+            return getTypeCode().toString();
+        }
+    }
+
+    /**
+     * A functional {@link Type} that has an optional list of argument {@link Type} and a return {@link Type}.
+     */
+    class Function implements Type {
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java:506]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java [hunk 10]
```diff
@@ -0,0 +1,1400 @@
...
+                                 .orElseGet(() -> field.getFieldType().toString());
+                     }).collect(Collectors.joining(", ")) + ")";
+        }
+
+        /**
+         * Creates a new erased {@link Record} type instance and returns it.
+         *
+         * @return a new erased {@link Record} type instance.
+         */
+        public static Record erased() {
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Type.java:783]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Typed.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
+/*
+ * Atom.java
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Typed.java:2]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Typed.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+     * @param formatter The formatter used to format the textual representation.
+     * @return a human-friendly textual representation of both the type-producing instance and its result set {@link Type}.
+     */
+    @Nonnull
+    String explain(@Nonnull final Formatter formatter);
+
+    /**
+     * Utility class for producing {@link Type} information of a given literal object.
+     */
+    class TypedLiteral implements Typed {
```
[the remark this answers is not in the document]
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Typed.java:68]: I think not, I removed it.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/QuantifiedColumnValue.java (hunk 1 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/QuantifiedColumnValue.java:156]: I think you mean we should be able to construct the record type information, right?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java (hunk 3 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java:204]: It was a mistake, fix it, thanks!

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/temp/TypeTest.java (hunk 4 above)
Comment (hatyo) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/temp/TypeTest.java:53]: :)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java (hunk 2 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/predicates/Value.java:83]: Alright, I renamed it accordingly.

**Review by normen662 — APPROVED** [review]

**Pull request merged**